File: haskell/secret-handshake/src/SecretHandshake.hs
module SecretHandshake (handshake) where

import Data.Bits (testBit)

handshake :: Int -> [String]
handshake n
  | testBit n 4 = reverse xs
  | otherwise = xs
  where
    xs =
      ["wink" | testBit n 0]
        ++ ["double blink" | testBit n 1]
        ++ ["close your eyes" | testBit n 2]
        ++ ["jump" | testBit n 3]

File: typescript/secret-handshake/secret-handshake.ts
export const commands = (n: number): string[] => {
  const toBinary = (n: number): string => n.toString(2)
  
  return toBinary(n).split('').reverse().reduce((acc: string[], curr: string, i: number) => {
    if (curr === '1') {
      switch (i) {
        case 0: return [...acc, 'wink']
        case 1: return [...acc, 'double blink']
        case 2: return [...acc, 'close your eyes']
        case 3: return [...acc, 'jump']
        case 4: return acc.reverse()
      }
    } 
    return acc
  }, [])
}

